### PR TITLE
Issue 5802: Optimized resource lookups when creating archives

### LINKED ISSUE #5802: Optimise the ArchiveBuilder
Currently, the archive builder takes a very long time.

For my test project, it takes 65-90 s, and yields a 14mb game.arcd file.

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java [hunk 1]
```diff
@@ -250,12 +250,15 @@ public void write(RandomAccessFile archiveIndex, RandomAccessFile archiveData, P
         // Write sorted entries to index file
         int entryOffset = (int) archiveIndex.getFilePointer();
         alignBuffer(archiveIndex, 4);
+
+        ByteBuffer indexBuffer = ByteBuffer.allocate(4 * 4 * entries.size());
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java:254]: Before: 1.5s
After: 0.15s

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java [hunk 2]
```diff
@@ -50,6 +52,7 @@
     private static final List<String> ENCRYPTED_EXTS = Arrays.asList("luac", "scriptc", "gui_scriptc", "render_scriptc");
 
     private List<ArchiveEntry> entries = new ArrayList<ArchiveEntry>();
+    private Set<String> lookup = new HashSet<String>(); // To see if a resource has already been added
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java:55]: Saved ~ 4s on this lookup test.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java [hunk 3]
```diff
@@ -398,28 +410,41 @@ public void addResourceEntry(String url, byte[] data, int flags) throws IOExcept
...
+            pathToOccurrances = new HashMap<>();
+
+            ResourceNode root = getRoot();
+            if (root != null) // for tests really
+                buildResourceOccurrancesMap(root);
         }
 
+        List<ArrayList<String>> result = new ArrayList<ArrayList<String>>();
+
+        List<ResourceNode> candidates = pathToOccurrances.get(filepath);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ManifestBuilder.java:444]: Saved 60+ s on this lookup table.

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java (hunk 2 above)
Comment (Jhonnyg) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/ArchiveBuilder.java:58]: 👏

**Review by Jhonnyg — DISMISSED** [review]
Looks good! 👍

**Pull request merged**